include/ginkgo/core/base/dense_cache.hpp

```diff
@@ -90,6 +100,68 @@ struct DenseCache {
 };
 
 
+/**
+ * Manages a workspace to give Dense Vector with different value_type. The
+ * workspace is buffered and reused internally to avoid repeated allocations.
+ * Copying an instance will only yield an empty object since copying the cached
+ * vector would not make sense. The stored object is always mutable, so the
+ * cache can be used in a const-context.
+ *
+ * @internal  The struct is present to wrap cache-like buffer storage that will
+ *            not be copied when the outer object gets copied.
+ */
+struct GenericDenseCache {
+    GenericDenseCache() = default;
+    ~GenericDenseCache() = default;
+    GenericDenseCache(const GenericDenseCache&);
+    GenericDenseCache(GenericDenseCache&&) noexcept;
+    GenericDenseCache& operator=(const GenericDenseCache&);
+    GenericDenseCache& operator=(GenericDenseCache&&) noexcept;
+    mutable array<char> workspace;
+
+    /**
+     * Pointer access to the underlying vector with specific type.
+     *
+     * @return  Pointer to the vector view.
+     */
+    template <typename ValueType>
+    std::shared_ptr<matrix::Dense<ValueType>> get(
+        std::shared_ptr<const Executor> exec, dim<2> size) const;
+};
+
+
+/**
+ * Manages a map to store Dense Scalar with different value_type by a
+ * user-specified value. The workspace is buffered and reused internally to
+ * avoid repeated allocations. Copying an instance will only yield an empty
+ * object since copying the cached vector would not make sense. The stored
+ * object is always mutable, so the cache can be used in a const-context.
+ *
+ * @internal  The struct is present to wrap cache-like buffer storage that will
+ *            not be copied when the outer object gets copied.
+ */
+struct ScalarCache {
+    ScalarCache(std::shared_ptr<const Executor> executor, double scalar_value);
+    ~ScalarCache() = default;
+    ScalarCache(const ScalarCache& other);
+    ScalarCache(ScalarCache&& other) noexcept;
+    ScalarCache& operator=(const ScalarCache& other);
+    ScalarCache& operator=(ScalarCache&& other) noexcept;
+    std::shared_ptr<const Executor> exec;
+    double value;
```

Comment:
can you make this private? Changing it will not invalidate the stored dense vectors, so perhaps some safety would be nice.

Reply:
I make them private now. I think there is no easy way to change the value now. Do you have something in mind for the safety check?

Reply:
Thanks, that is already enough.